core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.StatusCode;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+
+/**
+ * Factory for creating instances of the AccumuloProtocol.
+ * <p>
+ * This protocol includes a custom header to ensure compatibility between different versions of the
+ * protocol.
+ */
+public class AccumuloProtocolFactory extends TCompactProtocol.Factory {
+
+  private static final long serialVersionUID = 1L;
+
+  private final boolean isClient;
+
+  public static class AccumuloProtocol extends TCompactProtocol {
+
+    private static final int MAGIC_NUMBER = 0x41434355; // "ACCU" in ASCII
+    private static final byte PROTOCOL_VERSION = 1;
+    private static final boolean HEADER_HAS_TRACE = true;
+
+    private final boolean isClient;
+
+    private Span span = null;
+    private Scope scope = null;
+    private final Map<String,String> traceHeaders = new HashMap<>();
+
+    public AccumuloProtocol(TTransport transport, boolean isClient) {
+      super(transport);
+      this.isClient = isClient;
+    }
+
+    /**
+     * For client calls, add RPC span and write the validation header
+     */
+    @Override
+    public void writeMessageBegin(TMessage message) throws TException {
+      if (!this.isClient) {
+        super.writeMessageBegin(message);
+      } else {
+        span = TraceUtil.startClientRpcSpan(this.getClass(), message.name);
+        scope = span.makeCurrent();
+
+        try {
+          this.writeHeader();
+          super.writeMessageBegin(message);
+        } catch (TException e) {
+          if (span != null) {
+            span.recordException(e);
+            span.setAttribute("error.type", e.getClass().getCanonicalName());
+            span.setStatus(StatusCode.ERROR, e.getMessage());
+          }
+          if (scope != null) {
+            scope.close();
+          }
+          if (span != null) {
+            span.end();
+          }
+          throw e;
+        }
+      }
+    }
+
+    /**
+     * Writes the Accumulo protocol header containing version and identification info
+     */
+    private void writeHeader() throws TException {
+      super.writeI32(MAGIC_NUMBER);
+      super.writeByte(PROTOCOL_VERSION);
+
+      if (this.isClient && span != null && span.getSpanContext().isValid()) {
+        super.writeBool(HEADER_HAS_TRACE);
+        traceHeaders.clear();
+
+        W3CTraceContextPropagator.getInstance().inject(Context.current(), traceHeaders,
+            (headers, key, value) -> headers.put(key, value));
+
+        super.writeI16((short) traceHeaders.size());
```

Comment:
Should probably ensure traceHeaders.size() is less than Short.MAX_VALUE for safety, otherwise it could go negative.  However since this class extends TCompactProtocol could call `writeI32` instead which in TCompactProtocol will call `writeVarint32` which will make the integer as short as possible.   The impl of `writeI16` also calls `writeVarint32` so it does not seem there is much advantage to calling `writeI16` in this case.

Reply:
Fixed in 59a55ae. Seems like using `writeI32` makes the most sense here.